### PR TITLE
Task10 - initial commit

### PR DESCRIPTION
**Pull request by jb-user**
Task 10 - https://coherentsolutions.sharepoint.com/:w:/r/sites/training-center/_layouts/15/Doc.aspx?sourcedoc=%7B3A2E4457-28B3-49BF-AB13-F9B8AC1CAC2A%7D&file=Task10.docx&action=default&mobileredirect=true

### CODE REVIEW

File: src/main/java/com/coherent/api/training/TokenManager.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+public class TokenManager {
+
+    private static TokenManager instance;
+    private static String writeToken;
+    private static String readToken;
+
+    private TokenManager() {
+    }
+
+    public static TokenManager getInstance() {
```
Comment (shukal94) [inline src/main/java/com/coherent/api/training/TokenManager.java]: let's add synchronized to make your singleton thread-safe

File: src/main/java/com/coherent/api/training/MyHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+public class MyHttpClient {
+
+    private final TokenManager tokenManager;
+
+    public MyHttpClient() {
+        tokenManager = TokenManager.getInstance();
+    }
+
+    public String executeRequest(HttpRequestBase request) {
```
Comment (shukal94) [inline src/main/java/com/coherent/api/training/MyHttpClient.java]: probably we have to add capability to send unauthorized requests, so let's remove token setup from this method

File: src/main/java/com/coherent/api/training/TokenManager.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    public String getReadToken() {
+        if (readToken == null) {
+            readToken = getToken("read");
+        }
+        return readToken;
+    }
+
+    private String getToken(String scope) {
+        String token = null;
+        try (CloseableHttpClient httpClient = HttpClientBuilder.create().build()) {
```
Comment (shukal94) [inline src/main/java/com/coherent/api/training/TokenManager.java]: let's move urls, creds, params to properties file

File: src/main/java/com/coherent/api/training/TokenManager.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    public String getWriteToken() {
+        if (writeToken == null) {
+            writeToken = getToken("write");
+        }
+        return writeToken;
+    }
+
+    public String getReadToken() {
+        if (readToken == null) {
+            readToken = getToken("read");
```
Comment (shukal94) [inline src/main/java/com/coherent/api/training/TokenManager.java]: let's define enum with scope values

File: src/main/java/com/coherent/api/training/MyHttpClient.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package com.coherent.api.training;
+
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.*;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.util.EntityUtils;
+
+public class MyHttpClient {
```
Comment (shukal94) [inline src/main/java/com/coherent/api/training/MyHttpClient.java]: let's call it CustomHttpClient

File: src/test/java/com/coherent/api/training/ApiTests.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+import org.apache.http.client.methods.HttpGet;
+
+public class ApiTests {
+
+
+    // ToDo: Check was added temporary to execute GET request with read scope, remove
+    public static void main(String[] args) {
+        MyHttpClient httpClient = new MyHttpClient();
+        HttpGet getRequest = new HttpGet("http://localhost:8888/swagger-ui/");
```
Comment (shukal94) [inline src/test/java/com/coherent/api/training/ApiTests.java]: let's move url to properties file and add at least one validation to test

**Review by shukal94 — CHANGES_REQUESTED** [review]

File: src/test/java/com/coherent/api/training/ApiTests.java (hunk 6 above)
Comment (jb-user) [reply src/test/java/com/coherent/api/training/ApiTests.java]: code updated, please review

File: src/main/java/com/coherent/api/training/CustomHttpClient.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+    public static final  String BASE_URL = ConfigReader.getInstance().getProperty("base_url");
+
+    public static HttpResponse executeGet(String endpoint) {
+        return Request.getRequest(BASE_URL + endpoint)
+                .setBearerToken(AuthTokenManager.getInstance().getReadToken())
+                .executeRequest();
+    }
+
+    public static HttpResponse executePost(String endpoint) {
```
Comment (shukal94) [inline src/main/java/com/coherent/api/training/CustomHttpClient.java]: I think for post in future we'll need more parameters

**Review by shukal94 — CHANGES_REQUESTED** [review]

File: src/main/java/com/coherent/api/training/auth/AuthTokenResponse.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
+package com.coherent.api.training.auth;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class AuthTokenResponse {
```
Comment (shukal94) [inline src/main/java/com/coherent/api/training/auth/AuthTokenResponse.java]: move this class to model package

File: src/main/java/com/coherent/api/training/enums/Scope.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package com.coherent.api.training.enums;
```
Comment (shukal94) [inline src/main/java/com/coherent/api/training/enums/Scope.java]: it's not a good practice to have a separate enums package, it is model as well, so let's move all of your enums to model package

**Review by shukal94 — CHANGES_REQUESTED** [review]

File: src/main/java/com/coherent/api/training/enums/Scope.java (hunk 9 above)
Comment (jb-user) [reply src/main/java/com/coherent/api/training/enums/Scope.java]: fixed

**Review by shukal94 — APPROVED** [review]

**Pull request merged**